### PR TITLE
Issue/175 Template Engine for HTML UIs

### LINKED ISSUE #160: Allow Optional Elements in Questionnaire
Similar to optional Task.input elements.

### LINKED ISSUE #175: Template Engine for HTML UIs
Using a template engine like [Thymeleaf](https://www.thymeleaf.org) should improve the maintainability and robustness of the HTML generation code base.

### PR DESCRIPTION
**Pull request by hhund**
* Now using thymeleaf template engine to generate html
* Adds new UIs for some resources and search-sets
* Adds copy button for Task and QuestionnaireReponse html Input fields
* Adds insert placeholder button for boolean input fields
* Adds support for empty QuestionnaireReponse items and Task inputs using the FHIR data absent reason extension https://hl7.org/fhir/R4/extension-data-absent-reason.html
* Switching between html, json and xml views now pushes browser history entries
* Adds shortcuts (h, j, x) to switch between html, json and xml views. Shortcuts need to be used with browser specific modifier keys, see https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/accesskey
* Optional QuestionnaireResponse items
* UI mod feature allowing overrides of static resources via ui folder bind mound
* UI theme feature allowing configuration of a color indicator top page border for dev, test and prod deployment environments
* Initial BPE UI implementation listing deployed processes, active process instances, showing process BPMNs and allowing download of BPMNs.
* Revoked the authentication and authorization of BPE users, now using the same role config as the fhir
server. Currently only one role is defined "ADMIN". Access to BPE UI with the organization client certificate is not granted by default, but can be enabled by adding a role config the the "ADMIN" role for the certificate thumbprint.
* Added a bpe rev proxy docker image similar to the existing fhir rev proxy docker image.
* The jetty authentication / authorization code now also sets UserIdentity roles, enabling the use of jakarta.annotation.security.{DenyAll, PermitAll, RolesAllowed}
annotations on webservices.
* Added a new AuthenticationFilter (in common-auth) implementing the jakarta.ws.rs.container.ContainerRequestFilter interface to the bpe and fhir jersey applications to make sure (on the jersey layer) that access to all webservices is only granted to authenticated users.

closes #175
closes #160

### CODE REVIEW

File: dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceQuestionnaireResponse.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			return toItem(show, linkId, text, i.getAnswerFirstRep().getValue());
+		else
+			return new Item(show, linkId, null, text, null, null, null, null);
+	}
+
+	private Item toItem(boolean show, String id, String label, Type typedValue)
+	{
+		String fhirType = typedValue.getClass().getAnnotation(DatatypeDef.class).name();
+
+		if (typedValue instanceof BooleanType b)
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceQuestionnaireResponse.java:106]: Will be using switch expressions with pattern matching here after moving to java 21, wont fix for now.

File: dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceTask.java [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+			return type + ".reference";
+		else if (typedValue instanceof Reference r && r.hasIdentifier())
+			return type + ".identifier";
+		else
+			return type;
+	}
+
+	private String getStringValue(Type typedValue)
+	{
+		if (typedValue instanceof DecimalType d)
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceTask.java:269]: Will be using switch expressions with pattern matching here after moving to java 21, wont fix for now.

File: dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceTask.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+			logger.warn("Output parameter with {} value, not supported",
+					typedValue.getClass().getAnnotation(DatatypeDef.class).name());
+
+		return new OutputItem(UUID.randomUUID().toString(), type, label, labelTitle, stringValue, systemValueValue,
+				booleanValue, extension);
+	}
+
+	private String getHtmlInputType(Type typedValue)
+	{
+		if (typedValue instanceof BooleanType)
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/adapter/ResourceTask.java:223]: Will be using switch expressions with pattern matching here after moving to java 21, wont fix for now.

File: dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/webservice/impl/StaticResourcesServiceImpl.java [hunk 4]
```diff
@@ -72,29 +76,17 @@
...
-
-		private String mimeType(String fileName)
-		{
-			String[] parts = fileName.split("\\.");
-			return MIME_TYPE_BY_SUFFIX.get(parts[parts.length - 1]);
+			Path target = OVERRIDE_RESOURCE_FOLDER.resolve(fileName);
+			if (Files.isReadable(target))
+				return Files.newInputStream(target);
+			else
+				return StaticResourcesServiceImpl.class.getResourceAsStream("/static/" + fileName);
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/java/dev/dsf/fhir/webservice/impl/StaticResourcesServiceImpl.java:89]: false positive, `FILENAME_PATTERN` regex does not allow path traversal.

**Review by wetret — APPROVED** [review]

File: dsf-bpe/dsf-bpe-server/src/main/java/dev/dsf/bpe/webservice/AbstractService.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+import dev.dsf.bpe.ui.ThymeleafTemplateService;
+import dev.dsf.bpe.ui.ThymeleafTemplateService.MainValues;
+import dev.dsf.common.auth.conf.Identity;
+import jakarta.ws.rs.core.SecurityContext;
+import jakarta.ws.rs.core.StreamingOutput;
+
+public abstract class AbstractService implements InitializingBean
+{
+	@jakarta.ws.rs.core.Context
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-bpe/dsf-bpe-server/src/main/java/dev/dsf/bpe/webservice/AbstractService.java:18]: `org.thymeleaf.context.Context` is used in the same file and declared as an import.

File: dsf-bpe/dsf-bpe-server/src/main/resources/static/mod.css
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-bpe/dsf-bpe-server/src/main/resources/static/mod.css:1]: `mod.css` removed via 36edc1e

File: dsf-fhir/dsf-fhir-server/src/main/resources/static/mod.css
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/resources/static/mod.css:1]: `mod.css` removed via 36edc1e

**Pull request merged**